### ISSUE TITLE
Replace instrument PSAL with toolbox PSAL

### ISSUE THREAD

**Issue by FelicityMc** [aodn/imos-toolbox#515]
Is it possible to implement the ability to replace instrument derived PSAL with the Toolbox calculated variable - useful if an observed/consistent offset in CNDC  is applied to the data.

**Comment by FelicityMc** [comment]
Or alternatively the ability to keep/compare both the toolbox calculated PSAL and the instrument caclulated PSAL?

**Comment by ocehugo** [comment]
I got some questions before implementing this option. It may worth a quick discussion because this can change some aspects regarding data source/quality/distribution.

First assumptions/definitions:

1. We are talking about FV01 files.
2. I will call ```PSAL``` the actual variable written at FV01 level.
3. ```PSAL_INSTRUMENT``` is calculated from the instrument. It may be present or not. It's supposedly using the same ```CNDC_INSTRUMENT``` output from the sensor, but its value may not be completely reproducible since the vendor software is hidden/different/not available/etc.
4. ```PSAL_GSW_RAW``` is the psal calculated from TEOS-10 package using  ```CNDC_INSTRUMENT```, temp and pressure.
5. ```CNDC_ADJUSTED``` is CNDC corrected by a particular method X. So far, it's useful to generalise this as or a simple offset correction or a fancy function evaluation of several parameters.
6. ```PSAL_GSW_CORRECTED``` is the respective psal calculated with ```CNDC_ADJUSTED```. If there are several methods, then this (and CNDC) could be renamed/matched to them.

The current behaviour can be the following:

a. The instrument does not produce salinity estimates but got CTD vars.  ```PSAL -> PSAL_GSW_RAW```.

b. The instrument does produce salinity estimates. ```PSAL -> PSAL_INSTRUMENT```

c. After either b/a, and on the FV01 creation step, the operator can detect a problem in the original conductivity (```CNDC_INSTRUMENT```), and generate a fix for that (```CNDC_ADJUSTED```).

d. The operator then wishes to update ```PSAL -> PSAL_GSW_CORRECTED```.

The problem here is that we cant set easily ```PSAL -> PSAL_GSW_CORRECTED``` in either case (a/b). 

However, the big questions are:

I. Why do we not use a ```PSAL_GSW_CORRECTED``` all the time? I would assume that, if there is any possibility of a correction to ```CNDC_INSTRUMENT```, this should be done all the time, as part of the workflow, and nothing something that can be set aside.

II. Also, why one would trust more the instrument output variable than the one throw by TEOS-10?

III. The correction is easily applicable? Universal? What is the difference expected?

IV. Would operators find valid if we can provide a visual comparison before/after the correction? Correlation, bias? TS diagram? setting the metrics would be helpful and define if we can do it automatically, inform the operator, etc.

In any case, by current standards, we  should create new attributes before implementing this (e.g. ```psal: is_adjusted```, ```cndc:adjusted_offset```, ```cndc:adjustment_method:offset_removal```, etc.)

Requirements so far:
- [ ] A quick way to compare different CNDC estimates.
- [ ] Same for PSAL.
- [ ] The option to assign PSAL in FV01 (automatically or manually, TBD)
- [ ] Carry-over attributes to identify corrections (names/values TBD)

**Comment by petejan** [comment]
The answers to most of those questions is that the imos-toolbox is a QC tool. To verify that the data is correct, not to generate data from correct or incorrect data. 

The users of the tool are trying to generate there best estimate of the parameters, and document how they came to that data. 

So if a input file has PSAL `PSAL_INSTRUMENT` then being able to check that its the same as teos-10 'PSAL_CALCULATED' generated salinity is a useful thing to check, it checks that the temperature/conductivity/pressure are the correct ones from the instrument and that the instrument manufactures calculation is also correct, and there is no rounding errors (ok so now we have done some QC on those parameters).

If one later adjust one of the inputs to salinity, like conductivity, (and that's not possible with the manufactures tools) then one needs to re-calculate the PSAL to include it in the FV01 file.

The final file should only contain one PSAL (the best estimate of salinity that we have). 

If your going to add comment attributes then they should include why, not just what was done, like conductivity adjusted because the recovered calibration was different, or to align the salinity with CTD casts before/after.....

The FV00 file should be left as what was recorded by the instrument, if the instrument file never reported salinity, then don't report salinity in the FV00 file.

**Comment by ocehugo** [comment]
>The answers to most of those questions is that the imos-toolbox is a QC tool. To verify that the data is correct, not to generate data from correct or incorrect data.
>The users of the tool are trying to generate there best estimate of the parameters, and document how they came to that data.

I disagree. The toolbox is a QC **and** a metadata tool **plus** a data generation workflow. Otherwise, the toolbox would just contain QC-tests. The issue above is about reassignment and the inclusion of a correction on a variable (changing PSAL content from FV00 to FV01 with an "out-of-the-tree" correction).

>If one later adjust one of the inputs to salinity, like conductivity, (and that's not possible with the manufactures tools) then one needs to re-calculate the PSAL to include it in the FV01 file.

See, we are generating corrections and replacing content on the FV01.  Again - my point here is that this in-place change need proper registry **and** need to be reproducible. 
The content itself should be automatically generated. Otherwise, we have configuration/content drift, reprocessing manual steps/dependencies, reprocessing is problematic/slow, etc.

My point is, If the correction is not available in the toolbox, it's not easily traceable or (re)applicable. Hence, we should aim to allow this content change done in a procedural and reproducible way. Preferably in auto mode within the toolbox standard workflow. This would tick all the boxes. I'm not saying this would be automatically and "invisible" - it is quite the opposite and to help operator "qcing" and reproducibility.

>So if a input file has PSAL PSAL_INSTRUMENT then being able to check that its the same as teos-10 'PSAL_CALCULATED' generated salinity is a useful thing to check, it checks that the temperature/conductivity/pressure are the correct ones from the instrument and that the instrument manufactures calculation is also correct, and there is no rounding errors (ok so now we have done some QC on those parameters).

Agree, and I'm asking feedback (IV) to be able to implement a useful "index" for this decision making. Also, to implement an automatic way, I need to know the hierarchy and possible workflow scenarios.

>The final file should only contain one PSAL (the best estimate of salinity that we have).

Agree, but the content should be automatically generated from FV00 (or as close to an automatic procedure as possible -e.g. offset is stored and applied hierarchically).

>If your going to add comment attributes then they should include why, not just what was done, like conductivity adjusted because the recovered calibration was different, or to align the salinity with CTD casts before/after.....
>The FV00 file should be left as what was recorded by the instrument, if the instrument file never reported salinity, then don't report salinity in the FV00 file.

I was very clear that we are talking about FV01. The point of the discussion is to see if we favour implementing the actual procedure instead of just saying why.  We need to reproduce FV01 metadata & content.

**Comment by petejan** [comment]
So getting back to what Felicity asked, could you just give the salinityPP.m a configurable option to force overwriting of the existing salinity, much like the QC test have configurable options?

**Comment by ocehugo** [comment]
> So getting back to what Felicity asked, could you just give the salinityPP.m a configurable option to force overwriting of the existing salinity, much like the QC test have configurable options?

No need to worry about that. I'm just raising considerations/issues that will happen once the option is available.

@FelicityMc , may you provide me with an example file and the respective copy of the deployment database? I want to work on this specific case.

**Comment by FelicityMc** [comment]
Hi Hugo,
I am presently in the field and then away on leave for 4 weeks. Our internet is on satellite so very slow . I will ask Simon to please send on some example files from instruments we use:
SBE 37, SBE16, WQM
Regards
Felicity

**Comment by petejan** [comment]
Attached is a RAW SBE37SMP-ODO file, which can be processed using the SeaBird data processing tools available here, https://www.seabird.com/software-updates this tool allows you to process the RAW files into .CNV files selecting any variables to be included, this is where the salinity calculation is done

![image](https://user-images.githubusercontent.com/723918/62753585-a615cb00-baaf-11e9-9440-d2c8d4337904.png) 

![image](https://user-images.githubusercontent.com/723918/62753591-aca44280-baaf-11e9-8ef8-8641f9ab1ec0.png)


[SBE37-15969-30m.zip](https://github.com/aodn/imos-toolbox/files/3484699/SBE37-15969-30m.zip)

So you can create your own files with/without salinity for processing by the toolbox.
